File: src/prod/src/ServiceModel/reliability/failover/NodeDeactivationStatus.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace Common;
using namespace Reliability;
using namespace Reliability::NodeDeactivationStatus;

FABRIC_NODE_DEACTIVATION_STATUS Reliability::NodeDeactivationStatus::ToPublic(Enum deactivationStatus)
{
    switch (deactivationStatus)
    {
    case DeactivationSafetyCheckInProgress:
        return FABRIC_NODE_DEACTIVATION_STATUS_SAFETY_CHECK_IN_PROGRESS;
    case DeactivationSafetyCheckComplete:
        return FABRIC_NODE_DEACTIVATION_STATUS_SAFETY_CHECK_COMPLETE;
    case DeactivationComplete:
        return FABRIC_NODE_DEACTIVATION_STATUS_COMPLETED;
    default:
        return FABRIC_NODE_DEACTIVATION_STATUS_NONE;
    }
}

void NodeDeactivationStatus::WriteToTextWriter(TextWriter & w, Enum const& value)
{
    switch (value)
    {
        case DeactivationSafetyCheckInProgress:
            w << L"DeactivationSafetyCheckInProgress"; return;
        case DeactivationSafetyCheckComplete:
            w << L"DeactivationSafetyCheckComplete"; return;
        case DeactivationComplete:
            w << L"DeactivationComplete"; return;
        case ActivationInProgress:
            w << L"ActivationInProgress"; return;
        case None:
            w << L"None"; return;
        default:
            w << "NodeDeactivationStatus(" << static_cast<uint>(value) << ')'; return;
    }
}

NodeDeactivationStatus::Enum NodeDeactivationStatus::Merge(
    NodeDeactivationStatus::Enum const & left,
    NodeDeactivationStatus::Enum const & right)
{
    // This depends on the ordering of the enum values, which describe
    // the progression of a deactivation workflow
    //
    return std::min(left, right);
}

BEGIN_ENUM_STRUCTURED_TRACE(NodeDeactivationStatus);

ADD_ENUM_MAP_VALUE(NodeDeactivationStatus, DeactivationSafetyCheckInProgress);
ADD_ENUM_MAP_VALUE(NodeDeactivationStatus, DeactivationSafetyCheckComplete);
ADD_ENUM_MAP_VALUE(NodeDeactivationStatus, DeactivationComplete);
ADD_ENUM_MAP_VALUE(NodeDeactivationStatus, ActivationInProgress);
ADD_ENUM_MAP_VALUE(NodeDeactivationStatus, None)

END_ENUM_STRUCTURED_TRACE(NodeDeactivationStatus);

File: src/prod/src/managed/Api/src/System/Fabric/NodeDeactivationStatus.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric
{
    using System.Fabric.Interop;

    /// <summary>
    /// <para>Specified the status for a node deactivation task.</para>
    /// </summary>
    public enum NodeDeactivationStatus
    {
        /// <summary>
        /// <para>No status is associated with the task.</para>
        /// </summary>
        None = NativeTypes.FABRIC_NODE_DEACTIVATION_STATUS.FABRIC_NODE_DEACTIVATION_STATUS_NONE,
        /// <summary>
        /// <para>Safety checks are in progress for the task.</para>
        /// </summary>
        SafetyCheckInProgress = NativeTypes.FABRIC_NODE_DEACTIVATION_STATUS.FABRIC_NODE_DEACTIVATION_STATUS_SAFETY_CHECK_IN_PROGRESS,
        /// <summary>
        /// <para>All the safety checks have been completed for the task.</para>
        /// </summary>
        SafetyCheckComplete = NativeTypes.FABRIC_NODE_DEACTIVATION_STATUS.FABRIC_NODE_DEACTIVATION_STATUS_SAFETY_CHECK_COMPLETE,
        /// <summary>
        /// <para>The task is completed.</para>
        /// </summary>
        Completed = NativeTypes.FABRIC_NODE_DEACTIVATION_STATUS.FABRIC_NODE_DEACTIVATION_STATUS_COMPLETED
    }
}